### PR TITLE
synchronous: supprt more overlayfs layers

### PR DESCRIPTION
**Pull request by yuqitao**
Auto-detect longest common dir in lowerdir option and compact it if the
option size is hitting one page size. If does, Use chdir + CLONE to do
mount. see https://github.com/containerd/containerd/pull/2502

### CODE REVIEW

File: crates/shim/src/mount.rs [hunk 1]
```diff
@@ -223,6 +231,286 @@ lazy_static! {
...
+
+    let s = monitor_subscribe(Topic::All)?;
+    match unsafe { fork() } {
+        Ok(ForkResult::Parent { child, .. }) => {
+            let code: MountExitCode = wait_pid(i32::from(child), s).into();
+            code.into()
+        }
+        Ok(ForkResult::Child) => {
+            if let Some(workdir) = chdir {
+                env::set_current_dir(Path::new(&workdir)).unwrap_or_else(|_| {
```
Comment (Burning1020) [inline crates/shim/src/mount.rs:473]: I think it's no need to convert err code, you can print the detailed error log msg in child process and exit with none-zero code. We could check the log file for details if any error.

**Review by Burning1020 — COMMENTED** [review]
@fuweid It is also feasible to do chdir in the current process. I wonder how painful that would be.

File: crates/shim/src/error.rs [hunk 2]
```diff
@@ -114,6 +117,13 @@ macro_rules! io_error {
     };
 }
 
+#[macro_export]
+macro_rules! mount_process_error {
+    ($($args:tt)*) => {
+        Error::MountProcessError(format_args!($($args)*).to_string())
```
[the remark this answers is not in the document]
Comment (yuqitao) [reply crates/shim/src/error.rs]: updated

File: crates/shim/src/mount.rs [hunk 3]
```diff
@@ -223,6 +231,286 @@ lazy_static! {
...
+            _ => nix::errno::Errno::UnknownErrno,
+        }
+    }
+}
+
+impl From<MountExitCode> for Result<()> {
+    fn from(code: MountExitCode) -> Self {
+        match code {
+            MountExitCode::NixUnknownErr => Err(mount_process_error!(
+                "mount process exit unexpectly, exit code: {}",
```
[the remark this answers is not in the document]
Comment (yuqitao) [reply crates/shim/src/mount.rs]: updated

> unexpectedly

Done

File: crates/shim/src/mount.rs (hunk 1 above)
Comment (yuqitao) [reply crates/shim/src/mount.rs:472]: add detailed error log msg in child process.

**Comment by Burning1020** [comment]
LGTM

**Review by fuweid — APPROVED** [review]

**Review by mxpv — APPROVED** [review]

**Pull request merged**